File: clients/android/StellarChat/app/src/main/java/com/stellarmls/chat/blossom/BlossomClient.kt
package com.stellarmls.chat.blossom

import com.stellarmls.chat.crypto.KeyManager
import com.stellarmls.chat.crypto.NostrEventBuilder
import com.stellarmls.chat.model.toHex
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.security.MessageDigest
import java.util.concurrent.TimeUnit

/**
 * Client for Blossom (BL-01) content-addressable blob storage.
 * Files are addressed by SHA-256 hash of their content.
 * Upload authorization uses Nostr-signed kind 24242 events.
 */
object BlossomClient {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    /**
     * Upload an encrypted blob to Blossom server(s).
     * Returns the SHA-256 hex hash of the uploaded blob.
     */
    suspend fun upload(
        data: ByteArray,
        servers: List<String>,
        keyManager: KeyManager
    ): String {
        val hash = MessageDigest.getInstance("SHA-256").digest(data)
        val blobHash = hash.toHex()

        val authEvent = buildAuthEvent("upload", blobHash, keyManager)
        val authJson = authEvent.toJson().toString()
        val authToken = "Nostr ${android.util.Base64.encodeToString(
            authJson.toByteArray(), android.util.Base64.NO_WRAP)}"

        var lastError: Exception? = null
        for (server in servers) {
            try {
                val url = "${server.trimEnd('/')}/upload"
                val request = Request.Builder()
                    .url(url)
                    .put(data.toRequestBody("application/octet-stream".toMediaType()))
                    .header("Authorization", authToken)
                    .build()

                client.newCall(request).execute().use { response ->
                    if (response.isSuccessful) {
                        return blobHash
                    }
                    val reason = response.header("x-reason") ?: ""
                    val body = response.body?.string() ?: ""
                    val detail = reason.ifEmpty { body }
                    lastError = RuntimeException(
                        "Blossom upload to $server failed: HTTP ${response.code} $detail"
                    )
                }
            } catch (e: Exception) {
                lastError = e
            }
        }

        throw lastError ?: RuntimeException("Failed to upload to any Blossom server")
    }

    /**
     * Download an encrypted blob from Blossom server(s) by its SHA-256 hash.
     */
    suspend fun download(
        blobHash: String,
        servers: List<String>
    ): ByteArray {
        var lastError: Exception? = null
        for (server in servers) {
            try {
                val url = "${server.trimEnd('/')}/$blobHash"
                val request = Request.Builder().url(url).get().build()

                client.newCall(request).execute().use { response ->
                    if (response.isSuccessful) {
                        val data = response.body?.bytes()
                            ?: throw RuntimeException("Empty response body")
                        // Verify content hash matches
                        val downloadHash = MessageDigest.getInstance("SHA-256").digest(data)
                        val downloadHashHex = downloadHash.toHex()
                        if (downloadHashHex != blobHash) {
                            throw RuntimeException("Downloaded blob hash mismatch")
                        }
                        return data
                    }
                }
            } catch (e: Exception) {
                lastError = e
            }
        }

        throw lastError ?: RuntimeException("Failed to download from any Blossom server")
    }

    /**
     * Build a BL-01 authorization event (kind 24242).
     */
    private fun buildAuthEvent(
        action: String,
        blobHash: String,
        keyManager: KeyManager
    ): com.stellarmls.chat.crypto.NostrEvent {
        val expiration = (System.currentTimeMillis() / 1000) + 300 // 5 minutes

        val tags = listOf(
            listOf("t", action),
            listOf("x", blobHash),
            listOf("expiration", expiration.toString())
        )

        return NostrEventBuilder.build(
            kind = 24242,
            tags = tags,
            content = "Upload encrypted media",
            keyManager = keyManager
        )
    }
}

File: clients/ios/StellarChat/StellarChat/Models/BlossomClient.swift
import CryptoKit
import Foundation

/// Client for Blossom (BL-01) content-addressable blob storage.
/// Files are addressed by SHA-256 hash of their content.
/// Upload authorization uses Nostr-signed kind 24242 events.
enum BlossomClient {
    /// Extended session for large uploads (video). 300s request timeout.
    private static let extendedSession: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 300
        config.timeoutIntervalForResource = 600
        return URLSession(configuration: config)
    }()

    /// Upload an encrypted blob to Blossom server(s).
    /// Returns the SHA-256 hex hash of the uploaded blob.
    static func upload(
        _ data: Data,
        servers: [URL],
        keyManager: KeyManager
    ) async throws -> String {
        let hash = SHA256.hash(data: data)
        let blobHash = Data(hash).map { String(format: "%02x", $0) }.joined()

        let authEvent = try buildAuthEvent(
            action: "upload",
            blobHash: blobHash,
            keyManager: keyManager
        )
        let authJSON = try JSONSerialization.data(withJSONObject: authEvent.jsonObject)
        let authToken = "Nostr \(authJSON.base64EncodedString())"

        var lastError: Error?
        for server in servers {
            do {
                let url = server.appendingPathComponent("upload")
                var request = URLRequest(url: url)
                request.httpMethod = "PUT"
                request.httpBody = data
                request.setValue(authToken, forHTTPHeaderField: "Authorization")
                request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")

                let session = data.count > 2_000_000 ? extendedSession : URLSession.shared
                let (responseData, response) = try await session.data(for: request)
                if let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) {
                    return blobHash
                }
                let http = response as? HTTPURLResponse
                let statusCode = http?.statusCode ?? -1
                let reason = http?.value(forHTTPHeaderField: "x-reason") ?? ""
                let body = String(data: responseData, encoding: .utf8) ?? ""
                let detail = reason.isEmpty ? body : reason
                lastError = BlossomError.serverError(server: server.absoluteString, statusCode: statusCode, detail: detail)
            } catch {
                lastError = error
            }
        }

        throw lastError ?? BlossomError.uploadFailed
    }

    /// Download an encrypted blob from Blossom server(s) by its SHA-256 hash.
    static func download(
        blobHash: String,
        servers: [URL]
    ) async throws -> Data {
        var lastError: Error?
        for server in servers {
            do {
                let url = server.appendingPathComponent(blobHash)
                let (data, response) = try await URLSession.shared.data(from: url)
                if let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) {
                    // Verify content hash matches
                    let downloadHash = SHA256.hash(data: data)
                    let downloadHashHex = Data(downloadHash).map { String(format: "%02x", $0) }.joined()
                    guard downloadHashHex == blobHash else {
                        throw BlossomError.hashMismatch
                    }
                    return data
                }
            } catch {
                lastError = error
            }
        }

        throw lastError ?? BlossomError.downloadFailed
    }

    /// Build a BL-01 authorization event (kind 24242).
    private static func buildAuthEvent(
        action: String,
        blobHash: String,
        keyManager: KeyManager
    ) throws -> NostrEvent {
        let expiration = Int64(Date().timeIntervalSince1970) + 300 // 5 minutes

        let tags: [[String]] = [
            ["t", action],
            ["x", blobHash],
            ["expiration", String(expiration)],
        ]

        return try NostrEvent.build(
            kind: 24242,
            tags: tags,
            content: "Upload encrypted media",
            keyManager: keyManager
        )
    }
}

enum BlossomError: LocalizedError {
    case uploadFailed
    case downloadFailed
    case hashMismatch
    case serverError(server: String, statusCode: Int, detail: String)

    var errorDescription: String? {
        switch self {
        case .uploadFailed: return "Failed to upload to any Blossom server"
        case .downloadFailed: return "Failed to download from any Blossom server"
        case .hashMismatch: return "Downloaded blob hash does not match expected hash"
        case .serverError(let server, let statusCode, let detail):
            return "Blossom upload to \(server) failed: HTTP \(statusCode) \(detail)"
        }
    }
}
